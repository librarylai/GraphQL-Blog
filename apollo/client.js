import { useMemo } from 'react'
import { split, ApolloClient, InMemoryCache } from '@apollo/client'
import { createClient } from 'graphql-ws'
import { getMainDefinition } from '@apollo/client/utilities'


let apolloClient

function createIsomorphLink() {
  // server / statie 端 (SSR,SSG)
  if (typeof window === 'undefined') {
    const { SchemaLink } = require('@apollo/client/link/schema')
    const { schema } = require('./schema')
    const { getDB } = require('../mongodb')
    return new SchemaLink({
      schema,
      context: async () => {
        let db = await getDB()
        return {
          blogDB: db.collection('blog'),
          userDB: db.collection('user'),
        }
      },
    })
  } else {
    // Client 端 (CSR)
    const { HttpLink } = require('@apollo/client/link/http')
    const { GraphQLWsLink } = require('@apollo/client/link/subscriptions')
    const wsLink = new GraphQLWsLink(
      createClient({
        url: '/api/graphql',
      })
    )
    const httpLink = new HttpLink({
      uri: '/api/graphql',
      credentials: 'same-origin',
    })
    const splitLink = split(
      ({ query }) => {
        const definition = getMainDefinition(query)
        // 如果 operation 為 subscription 則 return true
        return definition.kind === 'OperationDefinition' && definition.operation === 'subscription'
      },
      wsLink, // 如果是 true 則為 wsLink
      httpLink // 如果是 false 則為 httpLink
    )
    return splitLink
  }
}

function createApolloClient() {
  return new ApolloClient({
    ssrMode: typeof window === 'undefined',
    link: createIsomorphLink(),
    cache: new InMemoryCache(),
  })
}

export function initializeApollo(initialState = null) {
  const _apolloClient = apolloClient ?? createApolloClient()

  // If your page has Next.js data fetching methods that use Apollo Client, the initial state
  // gets hydrated here
  if (initialState) {
    _apolloClient.cache.restore(initialState)
  }
  // For SSG and SSR always create a new Apollo Client
  if (typeof window === 'undefined') return _apolloClient
  // Create the Apollo Client once in the client
  if (!apolloClient) apolloClient = _apolloClient

  return _apolloClient
}

export function useApollo(initialState) {
  const store = useMemo(() => initializeApollo(initialState), [initialState])
  return store
}
