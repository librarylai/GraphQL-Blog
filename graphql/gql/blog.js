import gql from 'graphql-tag'

export const ALL_POST_QUERY = gql`
  query AllPostQuery {
    viewAllPost {
      id
      title
      content
    }
  }
`
export const ADD_POST_QUERY = gql`
  mutation AddPost($title: String, $content: String) {
    addPost(title: $title, content: $content) {
      id
      title
    }
  }
`
export const UPDATE_POST_QUERY = gql`
  mutation UpdatePost($postId: ID, $title: String, $content: String) {
    updatePost(postId: $postId, title: $title, content: $content){
      id
      title
    }
  }
`
export const DELETE_POST_QUERY = gql`
  mutation DeletePost($postId: ID) {
    deletePost(postId: $postId) {
      id
      title
    }
  }
`