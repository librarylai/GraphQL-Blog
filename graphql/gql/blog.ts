import gql from 'graphql-tag'

export const ALL_POST_QUERY = gql`
  query AllPostQuery {
    viewAllPost {
      id
      title
      content
      author {
        id
        name
      }
    }
  }
`
export const ADD_POST_QUERY = gql`
  mutation AddPost($title: String, $content: String, $authorId: ID) {
    addPost(title: $title, content: $content, authorId: $authorId) {
      id
      title
      content
      author {
        id
        name
      }
    }
  }
`
export const UPDATE_POST_QUERY = gql`
  mutation UpdatePost($postId: ID, $title: String, $content: String, $authorId: ID) {
    updatePost(postId: $postId, title: $title, content: $content, authorId: $authorId) {
      id
      title
      content
      author {
        id
        name
      }
    }
  }
`
export const DELETE_POST_QUERY = gql`
  mutation DeletePost($postId: ID) {
    deletePost(postId: $postId) {
      id
      title
      content
      author {
        id
        name
      }
    }
  }
`
// 新增評論
export const ADD_COMMENT = gql`
  mutation AddComment($postId: ID, $comment: String, $authorId: ID) {
    addComment(postId: $postId, comment: $comment, authorId: $authorId) {
      comment
    }
  }
`
