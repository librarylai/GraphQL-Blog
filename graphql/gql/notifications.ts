import gql from 'graphql-tag'

export const ALL_NOTIFICATIONS_QUERY = gql`
  query fetchNotifications {
    user {
      # 排序過後
      sortNotifications @client {
        id
        content {
          title
          subTitle
          createTime
          isRead
        }
      }
      # 原始資料
      notifications {
        id
        content {
          title
          subTitle
          createTime
          isRead
        }
      }
    }
  }
`

export const ALL_NOTIFICATIONS_TITLE_QUERY = gql`
  query fetchNotificationsTitle {
    user {
      notifications {
        id
        content {
          title
        }
      }
    }
  }
`

export const ALL_NOTIFICATIONS_SUBTITLE_QUERY = gql`
  query fetchNotificationsSubTitle {
    user {
      notifications {
        id
        content {
          subTitle
        }
      }
    }
  }
`

export const UPDATE_NOTIFICATION = gql`
  mutation updateNotification($id: ID) {
    updateNotification(notificationId: $id) {
      status
    }
  }
`
