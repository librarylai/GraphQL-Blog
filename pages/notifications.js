import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import Typography from '@mui/material/Typography'
import Grid from '@mui/material/Grid'
import Card from '@mui/material/Card'
import CardContent from '@mui/material/CardContent'
import { ALL_NOTIFICATIONS_QUERY } from '../graphql/gql/notifications'
import { useQuery } from '@apollo/client'
import moment from 'moment'
import InfiniteScroll from 'react-infinite-scroll-component'

const NotificationsContainer = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
`
const InfiniteWrapper = styled.div`
  width: 100%;
  height: 500px;
  overflow: auto;
  border: solid 1px #999;
  box-shadow: 1px 1px 5px #222 ;
`
const Font = styled.span`
  font-size: 24px;
  color: blue;
`

function Notifications(props) {
  const { data: notificationsData, fetchMore } = useQuery(ALL_NOTIFICATIONS_QUERY)
  const originNotifications = notificationsData?.user?.notifications
  const sortedNotifications = notificationsData?.user?.sortNotifications
  const fetchData = () => {
    if (fetchMore) {
      fetchMore({ query: ALL_NOTIFICATIONS_QUERY })
    }
  }
  // render 原始排序
  const renderOriginalView = () => {
    return (
      <InfiniteWrapper id='scrollableDiv'>
        <InfiniteScroll dataLength={originNotifications?.length ?? 0} next={fetchData} hasMore={true} scrollableTarget='scrollableDiv'>
          <Grid container>
            {originNotifications?.map((item) => {
              return (
                <Grid item key={item.id}>
                  <Card sx={{ width: '350px', height: '250px' }}>
                    <CardContent>
                      <Typography variant='h5' component='p'>
                        通知：{item?.content?.title}
                      </Typography>
                      <Typography variant='h6' component='p'>
                        副標題：{item?.content?.subTitle}
                      </Typography>
                      <Typography variant='h6' component='p'>
                        發文日期：{moment(Number(item?.content?.createTime)).format('YYYY/MM/DD')}
                      </Typography>
                    </CardContent>
                  </Card>
                </Grid>
              )
            })}
          </Grid>
        </InfiniteScroll>
      </InfiniteWrapper>
    )
  }
  // render 排序過後資料
  const renderSortedView = () => {
    return (
      <InfiniteWrapper id='scrollableSort'>
        <InfiniteScroll dataLength={originNotifications?.length ?? 0} next={fetchData} hasMore={true} scrollableTarget='scrollableSort'>
          <Grid container>
            {sortedNotifications?.map((item) => {
              return (
                <Grid item key={item.id}>
                  <Card sx={{ width: '350px', height: '250px' }}>
                    <CardContent>
                      <Typography variant='h5' component='p'>
                        通知：{item?.content?.title}
                      </Typography>
                      <Typography variant='h6' component='p'>
                        副標題：{item?.content?.subTitle}
                      </Typography>
                      <Typography variant='h6' component='p'>
                        發文日期：{moment(Number(item?.content?.createTime)).format('YYYY/MM/DD')}
                      </Typography>
                    </CardContent>
                  </Card>
                </Grid>
              )
            })}
          </Grid>
        </InfiniteScroll>
      </InfiniteWrapper>
    )
  }
  return (
    <NotificationsContainer>
      <Typography variant='h4' component='div'>
        通知頁面
      </Typography>
      <Grid container spacing={2} justifyContent={'center'}>
        <Grid item xs={6}>
          <Font>尚未排序</Font>
          {renderOriginalView()}
        </Grid>
        <Grid item xs={6}>
          <Font>排序過後</Font>
          {renderSortedView()}
        </Grid>
      </Grid>
    </NotificationsContainer>
  )
}

export default Notifications
