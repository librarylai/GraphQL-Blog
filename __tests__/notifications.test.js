import { render } from '@testing-library/react'
import { MockedProvider } from '@apollo/client/testing'
import Notifications from '../pages/notifications'
import { ALL_NOTIFICATIONS_QUERY } from '../graphql/gql/notifications'

function createNotificationData(title, subTitle, isRead, createTime) {
  return {
    content: {
      title,
      subTitle,
      isRead,
      createTime,
      __typename: 'Content',
    },
    __typename: 'Notification',
  }
}

// 用來 mock notification request and response
const mocks = [
  {
    request: {
      query: ALL_NOTIFICATIONS_QUERY,
      variables: {}, // 如果有需要帶入查詢參數，則要加入 variables
    },
    result: {
      data: {
        user: {
          notifications: [
            createNotificationData('library', '天氣真好', false, '2022/07/09'),
            createNotificationData('想買遊戲王卡', '青眼白龍勒～～', false, '2022/07/08'),
          ],
        },
      },
    },
  },
]

it('test notification page', () => {
  const { container } = render(
    <MockedProvider mocks={mocks}>
      <Notifications />
    </MockedProvider>
  )
  expect(container).toMatchSnapshot()
})
