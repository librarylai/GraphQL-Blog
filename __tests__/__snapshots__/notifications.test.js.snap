// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test notification page 1`] = `
<div>
  <div
    class="notifications__NotificationsContainer-sc-1x1x4ou-0 cWurVW"
  >
    <div
      class="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
    >
      通知頁面
    </div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-12n0uxm-MuiGrid-root"
    >
      <div
        class="notifications__LoadingWrapper-sc-1x1x4ou-4 kpeIWm"
      >
        Loading......
      </div>
    </div>
  </div>
</div>
`;
